evm/p2p/state.py

```diff
@@ -0,0 +1,187 @@
+            await asyncio.sleep(0.5)
+        peer = self.peer_pool.peers[random.randrange(0, len(self.peer_pool.peers))]
+        return cast(ETHPeer, peer)
+
+    async def stop(self):
+        await self.peer_pool.stop()
+
+    async def request_next_batch(self):
+        requests = self.scheduler.next_batch(MAX_STATE_FETCH)
+        if not len(requests):
+            # Although our run() loop frequently yields control to let our msg handler process
+            # received nodes (scheduling new requests), there may be cases when the pending nodes
+            # take a while to arrive thus causing the scheduler to run out of new requests for a
+            # while.
+            self.logger.debug("Scheduler queue is empty, not requesting any nodes")
+            return
+        self.logger.debug("Requesting %d trie nodes" % len(requests))
+        await self.request_nodes([request.node_key for request in requests])
+
+    async def request_nodes(self, node_keys):
+        peer = await self.get_random_peer()
+        now = time.time()
+        for node_key in node_keys:
+            self._pending_nodes[node_key] = now
+        peer.eth_proto.send_get_node_data(node_keys)
+
+    async def retry_timedout(self):
+        timed_out = []
+        now = time.time()
+        for node_key, req_time in list(self._pending_nodes.items()):
+            if now - req_time > self._reply_timeout:
+                timed_out.append(node_key)
+        if len(timed_out) == 0:
+            return
+        self.logger.debug("Re-requesting %d trie nodes" % len(timed_out))
+        await self.request_nodes(timed_out)
+
+    async def run(self):
+        asyncio.ensure_future(self.peer_pool.run())
+
+        self.logger.info("Starting state sync for root hash %s" % encode_hex(self.root_hash))
+        while self.scheduler.has_pending_requests:
+            # Request new nodes if we haven't reached the limit of pending nodes.
+            if len(self._pending_nodes) < self._max_pending:
+                await self.request_next_batch()
+
+            # Retry pending nodes that timed out.
+            if len(self._pending_nodes):
+                await self.retry_timedout()
+
+            if len(self._pending_nodes) > self._max_pending:
+                # Slow down if we've reached the limit of pending nodes.
+                self.logger.debug("Pending trie nodes limit reached, sleeping a bit")
+                await asyncio.sleep(0.3)
+            else:
+                # Yield control to ensure the Peer's msg_handler callback is called to process any
+                # nodes we may have received already. Otherwise we spin too fast and don't process
+                # received nodes often enough.
+                await asyncio.sleep(0)
+
+            self._maybe_report_progress()
+
+        self.logger.info("Finished state sync with root hash %s" % encode_hex(self.root_hash))
+
+    def _maybe_report_progress(self):
```

Comment:
Eventually, it seems like it would be more appropriate for this type of reporting to happen at an orchestration level higher up in the stack which is handling all the various long-running processes that need to be run.

Reply:
Yeah, that might make sense. Another alternative would be to handle it in the base class for asyncio services (#280), so subclasses just have to implement, say, a `report_progress()` method